File: ruby/binary_tree_postorder_traversal.rb
# Given a binary tree, return the postorder traversal of its nodes' values.

# For example:
# Given binary tree {1,#,2,3},
#    1
#     \
#      2
#     /
#    3
# return [3,2,1].

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}

require_relative 'src/tree'

class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

def postorder_traversal(root)
	postorder_list = []
	stack = []
	last_visited = nil

	while root || stack.size != 0
		if root
			stack << root
			root = root.left
		else
			root = stack.pop

			if root.right.nil? || root.right == last_visited
				postorder_list << root.val
				last_visited = root
				root = nil
			else
				stack << root
				root = root.right
			end
		end
	end

	postorder_list
end

File: javascript/binary_tree_postorder_traversal.js
// Given a binary tree, return the postorder traversal of its nodes' values.

// For example:
// Given binary tree {1,#,2,3},
//    1
//     \
//      2
//     /
//    3
// return [3,2,1].

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var treeModule = require('./src/tree.js');

var postorderTraversal = function(root) {
	if (root === null)
		return [];

	var order, stack, lastVisited;
	order = [];
	stack = [];
	lastVisited = null;

	while (root !== null || stack.length !== 0)
	{
		if (root !== null)
		{
			stack.push(root);
			root = root.left;
		}
		else
		{
			root = stack.pop();
			if (root.right === null || root.right === lastVisited)
			{
				order.push(root.val);
				lastVisited = root;
				root = null;
			}
			else
			{
				stack.push(root);
				root = root.right;				
			}
		}
	}

	return order;
};

var root = treeModule.createTree();
treeModule.preorderTraverse(root);
var order;
order = postorderTraversal(root);
console.log(order);